src/plonk/mod.rs

```diff
@@ -49,31 +48,45 @@ pub struct PlonkStructure<C: CurveAffine> {
     pub(crate) k: usize,
     pub(crate) selectors: Vec<Vec<bool>>,
     pub(crate) fixed_columns: Vec<Vec<C::ScalarExt>>,
-    pub(crate) num_advice_columns: usize,
-    /// total number of challenges
-    pub(crate) num_challenges: usize,
-    /// combined custom gates
-    pub(crate) gate: MultiPolynomial<C::ScalarExt>,
     /// concatenate selectors and num_fixed_columns together, then commit
     pub(crate) fixed_commitment: C,
+
+    pub(crate) num_advice_columns: usize,
+
+    /// We follow the special soundness protocol(SPS), section 3.1 in [Protostar](https://eprint.iacr.org/2023/620)
+    /// let k = num_challenges; when k > 0, we add extra verifier round, this is slightly different
+    /// from protostar paper.
+    /// see [`PlonkInstance.challenges`] for detail
+    pub(crate) num_challenges: usize,
+    /// specify the witness size of each prover round
+    pub(crate) round_sizes: Vec<usize>,
+
+    /// singla polynomial relation that combines custom gates and lookup relations
+    pub(crate) poly: MultiPolynomial<C::ScalarExt>,
     pub(crate) permutation_matrix: SparseMatrix<C::ScalarExt>,
-    pub(crate) lookups: Vec<lookup::Argument<C::ScalarExt>>,
+    pub(crate) lookup_arguments: Option<lookup::Arguments<C::ScalarExt>>,
 }
 
 #[derive(Clone, Debug)]
 pub struct PlonkInstance<C: CurveAffine> {
-    /// concatenate num_advice_columns together, then commit
-    pub(crate) W_commitment: C,
+    /// `W_commitments = round_sizes.len()`, see [`PlonkStructure.round_sizes`]
+    pub(crate) W_commitments: Vec<C>,
     /// inst = [X0, X1]
     pub(crate) instance: Vec<C::ScalarExt>,
-    /// contains challenges i.e. random number generated by verifier or Fiat-Shamir hash
+    /// challenges generated in special soundness protocol
+    /// we will have 0 ~ 3 challenges depending on different cases:
+    /// name them as r1, r2, r3.
+    /// r1: compress vector lookup, e.g. (a_1, a_2, a_3) -> a_1 + r1*a_2 + r1^2*a_3
+    /// r2: challenge to calculate h and g in log-derivative relation
+    /// r3: combine all custom gates (P_i) and lookup relations (L_i), e.g.:
+    /// (P_1, P_2, L_1, L_2) -> P_1 + r3*P_2 + r3^2*L_1 + r3^3*L_2
     pub(crate) challenges: Vec<C::ScalarExt>,
```

Comment:
We have many places with implicit size restrictions. I suggest for [discussion](https://github.com/snarkify/sirius/issues/72) to start using tinyvec crate, which allows to limit the maximum length. It would be hard enough to keep track of the potential invariants of such a large number of parameters and modify them, it would be better to do it with a compiler

Reply:
sounds good. After I finish lookup, I can replace vec to tiny vector for optimization.